distribution/lib/Standard/Database/0.0.0-dev/src/Internal/IR/Query.enso

```diff
@@ -104,3 +104,23 @@ type Query
        This will usually be a query of the form
        `DELETE FROM target WHERE (key_columns...) EXISTS (SELECT 1 FROM source WHERE source.key_columns = target.key_columns)`.
     Delete_Matching_Rows (target_table_name:Text) (source_table_name:Text) (key_columns : Vector Text)
+
+    traverse : (Any -> Any) -> Query
```

Comment:
We'll probably want a side-effecting version of this that does not rebuild the query.

Reply:
How would that be possible if our atoms are immutable?

I wouldn't be worried tbh, our queries tend to get large, but they are still very small in comparison to e.g. amount of data processed on Tables.

Reply:
Ah I think I now know what you meant above.

You meant that a `traverse` that does not return a new query but instead computes something, like the node count? And in such case, the query remains unchanged so it is wasteful to rebuild it? Fair.

Reply:
I think it could be helpful to document _somewhere_ how traverse works.

From what I see it traverses the IR calling the **same function** `f` on various IR types - and it's the responsibility of the function to detect if it wants to work on a given type or just return it unchanged.

I've grown used to having to provide separate callbacks for each type that is being traversed - such is a bit more 'type-safe' way to do this. I think your trick with using a single function and differentiating by argument type on call is good, but it would be great to describe it, as IMHO it is not that obvious.

Reply:
Done.
(I assume that I don't need PRIVATE on the methods of the private type `Let_Binder_Renamer` -- is that correct?)